Use `int32_t` type for most transaction size/weight values

hebasto opened the pull request:
From bitcoin/bitcoin#23957 which has been incorporated into this PR:
> A file-wide suppression is problematic because it will wave through future violations, potentially bugs.
> 
> Fix that by using per-statement casts.
> 
> This refactor doesn't change behavior because the now explicit casts were previously done implicitly.
> 
> Similar to commit 8b5a4de904b414fb3a818732cd0a2c90b91bc275

MarcoFalke commented:
Concept ACK. Could squash, given that the other pull was closed?

Things to look out for in review:

* On 32-bit this will potentially (minimally) reduce the number of txs that fit into the mempool
* types in cstdint aren't type safe and call sites will need to be reviewed as well

hebasto commented:
Updated 9a8a8adfff029f237a2934c2f9ec5e62644ea8e6 -> 39c82fa26093c7309f7b960959b5b395bba74a10 ([pr23962.01](https://github.com/hebasto/bitcoin/commits/pr23962.01) -> [pr23962.02](https://github.com/hebasto/bitcoin/commits/pr23962.02)).

> Could squash, given that the other pull was closed?

Commits have been reordered to keep changes focused.

shaavan reviewed (commented):
Concept ACK

I think it makes sense to use **int64_t** consistently for all transaction sizes.

I have yet to thoroughly review the file to check if all the transaction size variables are correctly covered in this PR; I still find some that could use a proper type assignment.

1. [Here](https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L214). The type of `entry_size` can be changed to **int64_t**. This will lead to making appropriate changes at the other places in the file.
2. [Here](https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L224). Type of `totalSizeWithAncestors` could be changed.

Btw I was curious about one thing. In [line 232](https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L232) in this file, there is implicit typecasting occurring:
```
totalSizeWithAncestors += stageit->GetTxSize();
```
because type of `totalSizeWithAncestors` is **size_t** and return type of `GetTxSize()` is  changed to **int64_t** in this PR (line 114) . Shouldn't it cause implicit-conversion-warning for this.

glozow reviewed (commented):
Concept ACK to unifying types for transaction sizes. But if we're going to do this, why use a signed type? We don't have negative sizes, do we?

MarcoFalke commented:
> Concept ACK to unifying types for transaction sizes. But if we're going to do this, why use a signed type? We don't have negative sizes, do we?

The rule of thumb is to use signed types for anything that involves arithmetic operations (addition or subtraction). Obviously the standard library doesn't make that easy with `size_t` being unsigned.

hebasto commented:
@glozow 
> Concept ACK to unifying types for transaction sizes. But if we're going to do this, why use a signed type? We don't have negative sizes, do we?

Also see [Signed and Unsigned Types in Interfaces](https://www.aristeia.com/Papers/C++ReportColumns/sep95.pdf).

luke-jr commented:
Concept NACK. This wastes memory for no benefit - we should probably go to `int32_t` instead.

Transactions should never approach 2 GB, and it doesn't make sense to calculate a sum that large either (the largest package that could fit in a block is 4 MB).

If using `int32_t` produces warnings somewhere, those warnings are probably real issues that should be fixed.

luke-jr reviewed (changes requested):
Concept NACK [edit: to 64-bit sizes, which this PR no longer is]. This wastes memory for no benefit - we should probably go to `int32_t` instead.

Transactions should never approach 2 GB, and it doesn't make sense to calculate a sum that large either (the largest package that could fit in a block is 4 MB).

If using `int32_t` produces warnings somewhere, those warnings are probably real issues that should be fixed.

hebasto commented:
Updated 39c82fa26093c7309f7b960959b5b395bba74a10 -> 295a4f08728213270a0958680594c7548832ccfa ([pr23962.02](https://github.com/hebasto/bitcoin/commits/pr23962.02) -> [pr23962.03](https://github.com/hebasto/bitcoin/commits/pr23962.03)):

- rebased
- addressed @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/23962#pullrequestreview-850055431):
> This wastes memory for no benefit - we should probably go to `nt32_t` instead.

---

@shaavan 
> I think it makes sense to use int64_t consistently for all transaction sizes.

You are right. But such a pr could be unmanageable in size. So here, in the first commit, the minimal set of changes included which makes the second commit possible and concise. Other changes left for follow ups.

hebasto commented:
Rebased 295a4f08728213270a0958680594c7548832ccfa -> 960f996449694fbf67ed0e70f0c1a81323cf151c ([pr23962.03](https://github.com/hebasto/bitcoin/commits/pr23962.03) -> [pr23962.04](https://github.com/hebasto/bitcoin/commits/pr23962.04)) due to the conflict with #21464.

shaavan reviewed (commented):
Changes since my review:
- Replaced **int64_t** with **int32_t** for the transaction variables. This is done for the reason suggested in [this](https://github.com/bitcoin/bitcoin/pull/23962#pullrequestreview-850055431) comment, with which I agree.
- Rebased over the current master.

I observed the following implicit conversion warning:

![Screenshot from 2022-01-25 20-06-15](https://user-images.githubusercontent.com/85434418/150997503-dc7d7eac-09ed-4e33-93ec-08238879030b.png)

Ideally, this needs to be addressed by changing the `ancestor_size_limit` to **int32_t** and making other subsequent changes. But for this PR, we can go with at-place manual conversion, if that's possible.

@hebasto
> You are right. But such a pr could be unmanageable in size. So here, in the first commit, the minimal set of changes included which makes the second commit possible and concise. Other changes left for follow ups.

That makes sense. Let's take up changing small sections of transaction_size variables one by one in follow-ups.

hebasto commented:
> I observed the following implicit conversion warning:

Must be fixed now.

shaavan reviewed (commented):
Changes since my last review:

- Argument type of `ancestor_size_limit` in **CTxMemPool::UpdateForDescendants** function changed from **uint64_t** -> **int64_t**

I tried experimenting with converting the type of `ancestor_size_limit` to **int32_t**, and it compiled successfully without warnings.
So what do you think about converting this argument type to `int32_t` instead of `int64_t`. This would be in line with our end goal to do so for all transaction size/weight variables?

kiminuo commented:
@hebasto I wonder what you think about modifying https://github.com/bitcoin/bitcoin/blob/a41976ab770d4453ccb043f72e48a6eb15a7106a/src/policy/feerate.cpp#L12

to 

```cpp
 CFeeRate::CFeeRate(const CAmount& nFeePaid, int32_t num_bytes) 
```

Not necessarily in this PR though.

This change would help me in #21422 (https://github.com/bitcoin/bitcoin/pull/21422/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR1687) to remove the need for casting.

I don't know if it is a good idea. Just asking.

hebasto commented:
Rebased 2a4cd379be4ed72ea9007f757f189c9335949aa7 -> 482843ec92674c7da1cc57800a910d232c9e70bc ([pr23962.05](https://github.com/hebasto/bitcoin/commits/pr23962.05) -> [pr23962.06](https://github.com/hebasto/bitcoin/commits/pr23962.06)).

@kiminuo 
> I wonder what you think about modifying
> ...
> Not necessarily in this PR though.

I'd like to keep this PR with changes required for removing of suppression from `test/sanitizer_suppressions/ubsan`.

hebasto commented:
Rebased 2a4cd379be4ed72ea9007f757f189c9335949aa7 -> 7dd7d2fb6bee51cb245069cf5d3a2ebe48d95fa5 ([pr23962.05](https://github.com/hebasto/bitcoin/commits/pr23962.05) -> [pr23962.06](https://github.com/hebasto/bitcoin/commits/pr23962.06)).

hebasto commented:
Rebased 7dd7d2fb6bee51cb245069cf5d3a2ebe48d95fa5 -> eb64c286e97c10109ff378834f649e1f379978ac ([pr23962.06](https://github.com/hebasto/bitcoin/commits/pr23962.06) -> [pr23962.07](https://github.com/hebasto/bitcoin/commits/pr23962.07)).

shaavan reviewed (commented):
Changes since my last review:

- PR rebased on master

**Suggestion:**

I want to propose one suggestion.
- The variables named `modifySize` deals with transaction size and are added with transaction size variables on multiple occasions. For example:
    
    ` nSizeWithDescendants += modifySize;`
    
- However, its type is kept as `int64_t`.
- This might be a deliberate choice based on the maximum possible size of integer that maybe get stored in it. But in case it's not, I would suggest changing its type to `int32_t`.

- I was able to successfully compile this PR with the following related diff:

```diff
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -25,7 +25,7 @@
 // Helpers for modifying CTxMemPool::mapTx, which is a boost multi_index.
 struct update_descendant_state
 {
-    update_descendant_state(int64_t _modifySize, CAmount _modifyFee, int64_t _modifyCount) :
+    update_descendant_state(int32_t _modifySize, CAmount _modifyFee, int64_t _modifyCount) :
         modifySize(_modifySize), modifyFee(_modifyFee), modifyCount(_modifyCount)
     {}
 
@@ -33,14 +33,14 @@ struct update_descendant_state
         { e.UpdateDescendantState(modifySize, modifyFee, modifyCount); }
 
     private:
-        int64_t modifySize;
+        int32_t modifySize;
         CAmount modifyFee;
         int64_t modifyCount;
 };
 
 struct update_ancestor_state
 {
-    update_ancestor_state(int64_t _modifySize, CAmount _modifyFee, int64_t _modifyCount, int64_t _modifySigOpsCost) :
+    update_ancestor_state(int32_t _modifySize, CAmount _modifyFee, int64_t _modifyCount, int64_t _modifySigOpsCost) :
         modifySize(_modifySize), modifyFee(_modifyFee), modifyCount(_modifyCount), modifySigOpsCost(_modifySigOpsCost)
     {}
 
@@ -48,7 +48,7 @@ struct update_ancestor_state
         { e.UpdateAncestorState(modifySize, modifyFee, modifyCount, modifySigOpsCost); }
 
     private:
-        int64_t modifySize;
+        int32_t modifySize;
         CAmount modifyFee;
         int64_t modifyCount;
         int64_t modifySigOpsCost;
@@ -134,7 +134,7 @@ void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap& cachedDescendan
     }
     // descendants now contains all in-mempool descendants of updateIt.
     // Update and add to cached descendant map
-    int64_t modifySize = 0;
+    int32_t modifySize = 0;
     CAmount modifyFee = 0;
     int64_t modifyCount = 0;
     for (const CTxMemPoolEntry& descendant : descendants) {
@@ -387,7 +387,7 @@ void CTxMemPool::UpdateForRemoveFromMempool(const setEntries &entriesToRemove, b
             setEntries setDescendants;
             CalculateDescendants(removeIt, setDescendants);
             setDescendants.erase(removeIt); // don't update state for self
-            int64_t modifySize = -((int64_t)removeIt->GetTxSize());
+            int32_t modifySize = -removeIt->GetTxSize();
             CAmount modifyFee = -removeIt->GetModifiedFee();
             int modifySigOps = -removeIt->GetSigOpCost();
             for (txiter dit : setDescendants) {
@@ -431,7 +431,7 @@ void CTxMemPool::UpdateForRemoveFromMempool(const setEntries &entriesToRemove, b
     }
 }
 
-void CTxMemPoolEntry::UpdateDescendantState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount)
+void CTxMemPoolEntry::UpdateDescendantState(int32_t modifySize, CAmount modifyFee, int64_t modifyCount)
 {
     nSizeWithDescendants += modifySize;
     assert(nSizeWithDescendants > 0);
@@ -440,7 +440,7 @@ void CTxMemPoolEntry::UpdateDescendantState(int64_t modifySize, CAmount modifyFe
     assert(int64_t(nCountWithDescendants) > 0);
 }
 
-void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps)
+void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps)
 {
     nSizeWithAncestors += modifySize;
     assert(nSizeWithAncestors > 0);
diff --git a/src/txmempool.h b/src/txmempool.h
index 7ba9fbefd..fe52995fc 100644
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -136,9 +136,9 @@ public:
     const LockPoints& GetLockPoints() const { return lockPoints; }
 
     // Adjusts the descendant state.
-    void UpdateDescendantState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount);
+    void UpdateDescendantState(int32_t modifySize, CAmount modifyFee, int64_t modifyCount);
     // Adjusts the ancestor state
-    void UpdateAncestorState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps);
+    void UpdateAncestorState(int32_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps);
     // Updates the fee delta used for mining priority score, and the
     // modified fees with descendants/ancestors.
     void UpdateFeeDelta(CAmount newFeeDelta);
```

hebasto commented:
Updated eb64c286e97c10109ff378834f649e1f379978ac -> 03300afb2fdb31a0516cb94dc253e985c35dbce2 ([pr23962.07](https://github.com/hebasto/bitcoin/commits/pr23962.07) -> [pr23962.08](https://github.com/hebasto/bitcoin/commits/pr23962.08)):

- rebased
- addressed @shaavan's [comment](https://github.com/bitcoin/bitcoin/pull/23962#pullrequestreview-1013575887)

stickies-v commented:
Concept ACK. Using 32 bits seems appropriate for sizes of transactions and transaction sets. Under the 100kb standardness rule, an unconfirmed transaction cannot exceed 400,000WU, which means this should never overflow for sets <= `(2^31)/400,000 = 5,368` (unconfirmed) transactions which seems sufficient given the current limitations on ancestors and descendants, with room to grow.

Moving to signed integers seems like a worthwhile trade-off: half the range, but increased visibility into unexpected under/overflows.

I'll do a full review once it's rebased.

hebasto commented:
Rebased 03300afb2fdb31a0516cb94dc253e985c35dbce2 -> f63e29016d97ac78876d683df7675fc8f57d9465 ([pr23962.08](https://github.com/hebasto/bitcoin/commits/pr23962.08) -> [pr23962.09](https://github.com/hebasto/bitcoin/commits/pr23962.09)).

hebasto commented:
Rebased f63e29016d97ac78876d683df7675fc8f57d9465 -> 316a53a838be66bc15354d96a0376c9ffdbac61f ([pr23962.09](https://github.com/hebasto/bitcoin/commits/pr23962.09) -> [pr23962.10](https://github.com/hebasto/bitcoin/commits/pr23962.10)).

hebasto commented:
Rebased 316a53a838be66bc15354d96a0376c9ffdbac61f -> 50e872ba71726488158097a7085c7c99ded1c777 ([pr23962.10](https://github.com/hebasto/bitcoin/commits/pr23962.10) -> [pr23962.11](https://github.com/hebasto/bitcoin/commits/pr23962.11)) due to the conflict with #26609.

Xekyo commented:
ACK 93daff4b4577a07994b57218df9bb83bed7bf743

hebasto commented:
> ISTM we should also change `GetBlockWeight`, `GetTransactionInputWeight`, `GetVirtualTransactionSize`, and `GetVirtualTransactionInputSize` to also use `int32_t`?

Agree. But I'd rather leave this PR as is, because suggested changes does look trivial and could be left for follow ups. For example, `GetVirtualTransactionSize` involves arithmetic operations with `int64_t nSigOpCost`, and additional changes/checks could be required.

achow101 commented:
ACK 93daff4b4577a07994b57218df9bb83bed7bf743

hebasto commented:
Friendly ping @glozow :)

ryanofsky commented on src/txmempool.cpp:
```diff
@@ -367,7 +367,7 @@ void CTxMemPoolEntry::UpdateDescendantState(int32_t modifySize, CAmount modifyFe
     assert(nSizeWithDescendants > 0);
     nModFeesWithDescendants = SaturatingAdd(nModFeesWithDescendants, modifyFee);
-    nCountWithDescendants += modifyCount;
+    nCountWithDescendants += uint64_t(modifyCount);
     assert(int64_t(nCountWithDescendants) > 0);
```
In commit "Remove txmempool implicit-integer-sign-change sanitizer suppressions" (93daff4b4577a07994b57218df9bb83bed7bf743)

IIUC, the code here casting `modifyCount` to be unsigned and adding it even though `modifyCount` might be negative is safe because the standard guarantees that:

1. Casting `modifyCount` to `uint64_t` produces an equivalent value mod 2<sup>64</sup>
2. Adding two `uint64_t` values is performed mod 2<sup>64</sup>

Also, the new code is equivalent to previous code because previous code was just doing the same conversion implicitly.

However, it seems like it be nice to follow up by making `nCountWithDescendants` signed so the two casts could be dropped. So:

```c++
nCountWithDescendants += uint64_t(modifyCount);
assert(int64_t(nCountWithDescendants) > 0);
```

could be replaced with:

```c++
nCountWithDescendants += modifyCount;
assert(nCountWithDescendants > 0);
```

hebasto replied:
@ryanofsky 

I agree with the suggested changes. In addition, it seems reasonable to combine them with making signed `nCountWithAncestors` and return types of the related getter methods as well. I'd leave such changes for a follow up. Or do you prefer to incorporate them in here?

ryanofsky replied:
Either way seems fine. I suggested it for a followup because it isn't directly related to using a consistent type for transaction sizes. If you will make a follow up PR, you could consider dropping the second commit from this PR to avoid churn, since the second commit 93daff4b4577a07994b57218df9bb83bed7bf743 also isn't directly related to tx sizes. But whatever you want to do is fine.

ryanofsky reviewed (approved):
Code review ACK 93daff4b4577a07994b57218df9bb83bed7bf743

ryanofsky commented:
Following patch fixes the silent merge conflict for me after a rebase:

```diff
diff --git a/src/test/miniminer_tests.cpp b/src/test/miniminer_tests.cpp
index 371147b0e23c..b26f7dafe342 100644
--- a/src/test/miniminer_tests.cpp
+++ b/src/test/miniminer_tests.cpp
@@ -137,7 +137,7 @@ BOOST_FIXTURE_TEST_CASE(miniminer_1p1c, TestChain100Setup)
 
     std::vector<CTransactionRef> all_transactions{tx1, tx2, tx3, tx4, tx5, tx6, tx7, tx8};
     struct TxDimensions {
-        size_t vsize; CAmount mod_fee; CFeeRate feerate;
+        int32_t vsize; CAmount mod_fee; CFeeRate feerate;
     };
     std::map<uint256, TxDimensions> tx_dims;
     for (const auto& tx : all_transactions) {

```

hebasto commented:
> Silent merge conflict (with `-werror`)

Rebased. Sorry for not addressing comments promptly.

> Following patch fixes the silent merge conflict for me after a rebase:

@ryanofsky Thanks! Applied.

hebasto committed:
Use `int32_t` type for most transaction size/weight values

This change gets rid of a few casts and makes the following commit diff
smaller.

hebasto committed:
Remove txmempool implicit-integer-sign-change sanitizer suppressions

Xekyo commented on src/kernel/mempool_entry.h:
```diff
@@ -152,13 +154,13 @@ class CTxMemPoolEntry
 
     uint64_t GetCountWithDescendants() const { return nCountWithDescendants; }
-    uint64_t GetSizeWithDescendants() const { return nSizeWithDescendants; }
+    int64_t GetSizeWithDescendants() const { return nSizeWithDescendants; }
```
Re-reviewing this basically from scratch, I stumble here. `int32_t` should be big enough to represent the weight of more than 5,000 max standard weight transactions. Are we actually bumping into overflow issues with `int32_t` somewhere for size with descendants?

I see that this was brought up before, so I assume it was left a 64-bit value to limit the scope of this PR. Is that right?

hebasto replied:
> Are we actually bumping into overflow issues with `int32_t` somewhere for size with descendants?

I don't think so.

However, UBSan raises "signed-integer-overflow" warnings about `nSizeWithAncestors += modifySize;` etc.

Xekyo commented on src/kernel/mempool_entry.h:
```diff
@@ -152,13 +154,13 @@ class CTxMemPoolEntry
 
     uint64_t GetCountWithAncestors() const { return nCountWithAncestors; }
-    uint64_t GetSizeWithAncestors() const { return nSizeWithAncestors; }
+    int64_t GetSizeWithAncestors() const { return nSizeWithAncestors; }
```
As above

murchandamus reviewed (commented):
codereview ACK 3ef756a5b558a1dd2fcb93bc0d4237707aa04f3f

ryanofsky reviewed (approved):
Code review ACK 3ef756a5b558a1dd2fcb93bc0d4237707aa04f3f. Since last review, just rebased with more type changes in test and tracing code

achow101 commented:
ACK 3ef756a5b558a1dd2fcb93bc0d4237707aa04f3f

achow101 commented:
Something that just occurred to me: do/should we consider tracepoints to be a stable api? Since this changes the types for some of the tracepoints, this would ostensibly break any third party programs that rely on those tracepoints.

hebasto commented:
The silent conflict with the https://github.com/bitcoin/bitcoin/pull/26531 has been resolved in the recent push. As this PR touches tracepoints now, @0xB10C @virtu maybe you want to look into it?

> Something that just occurred to me: do/should we consider tracepoints to be a stable api?

No, we shouldn't. I think that classes like `CTxMemPoolEntry` and their public access method return types should be considered as mempool implementation details for all purposes. Maybe tracepoint docs need an explicit note about that?

0xB10C commented:
ACK 3ef756a5b558a1dd2fcb93bc0d4237707aa04f3f. I've focused my testing and code review on the tracepoint related changes. The docs, the test, and the mempool_monitor.py demo script are updated. I ran the `interface_usdt_mempool.py` test and the `mempool_monitor.py` script. The `mempool_monitor.py` output looks correct.

>> Something that just occurred to me: do/should we consider tracepoints to be a stable api?
>
> No, we shouldn't. I think that classes like CTxMemPoolEntry and their public access method return types should be considered as mempool implementation details for all purposes.

The goal of the tracepoints is to expose implementation details. This means, if an implementation detail changes, the tracepoint interface can change too. As an extreme example: If we'd remove the mempool, we'd also drop the mempool tracepoints. Third party programs have to adapt in this case.

hebasto commented:
> This causes warnings when compiling for 32 bit?

The relevant CI tasks logs are clean. How to reproduce it?

glozow commented:
> This causes warnings when compiling for 32 bit?

@hebasto did you end up looking into this?

hebasto commented:
> This causes warnings when compiling for 32 bit?

> > This causes warnings when compiling for 32 bit?
> 
> @hebasto did you end up looking into this?

I apologise for forgetting to mention that #28059 addresses that warnings.
